Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c

```diff
@@ -0,0 +1,878 @@
+                                            IMX_VARIABLE_SMBIOS_MANUFACTURER,
+                                            IMX_VARIABLE_SMBIOS_PRODUCT_NAME,
+                                            IMX_VARIABLE_SMBIOS_SKU,
+                                            IMX_VARIABLE_SMBIOS_FAMILY,
+                                            IMX_VARIABLE_SMBIOS_BASEBOARD_PRODUCT};
+
+STATIC SMBIOS_OVERRIDE_NODE *mSmbiosOverrideListHead = NULL;
+
+/**
+
+  Retrieves the next Key/Value pair from the given INI buffer.
+
+  This routine scans through the provided buffer starting at the given offset.
+  The first character is expected to be the start of the Key. The routine
+  iterates through the buffer until it finds "=" character, which signifies
+  the end of the Key. The process repeats for the Value portion and ends
+  when a newline is found.
+
+  At any point if a failure is detected, the routine bails with an error.
+  If Key and Value are successfully found, allocate and return Unicode strings
+  containing Key/Value and return these to the caller.
+
+  Note: the caller is responsible for freeing these strings.
+
+  @param[in]      Buffer        Input INI buffer with ASCII characters.
+  @param[in out]  Offset        Pointer to starting offset into buffer. When
+                                routine is finished, this pointer is updated
+                                to hold the starting offset of the next INI
+                                key.
+  @param[in]      BufferSize    Total size of the input buffer in bytes.
+  @param[out]     Key           Pointer to Unicode string containing Key
+  @param[out]     Value         Pointer to Unicode string containing Value
+
+  @retval EFI_SUCCESS             
+  @retval EFI_INVALID_PARAMETER   Bad parameter passed into this function
+
+**/
+EFI_STATUS
+EFIAPI
+GetNextIniKeyValuePair (
+  IN  CHAR8       *Buffer,
+  IN OUT UINTN    *Offset,
+  IN  UINTN       BufferSize,
+  OUT CHAR16      **Key,
+  OUT CHAR16      **Value
+  )
+{
+  UINTN             CurrentOffset;
+  UINTN             KeyLength;
+  CHAR8             *KeyStart;
+  INI_PARSER_STATE  State;
+  EFI_STATUS        Status;
+  CHAR8             *TempAsciiKey;
+  CHAR8             *TempAsciiValue;
+  CHAR16            *TempUnicodeKey;
+  CHAR16            *TempUnicodeValue;
+  UINTN             ValueLength;
+  CHAR8             *ValueStart;
+
+  if ((Buffer == NULL) || (Offset == NULL) ||
+      (Key == NULL) || (Value == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  CurrentOffset = *Offset;
+  KeyLength = 0;
+  ValueLength = 0;
+  Status = EFI_NOT_FOUND;
+  State = StateReadKey;
+  KeyStart = NULL;
+  ValueStart = NULL;
+  TempAsciiKey = NULL;
+  TempAsciiValue = NULL;
+  TempUnicodeKey = NULL;
+  TempUnicodeValue = NULL;
+
+  //
+  // Loop until end of file or we find a valid key/value pair
+  //
+  while (CurrentOffset < BufferSize) {
+    if (State == StateValueDone) {
```

Comment:
Corner case: we don't return the final key/value pair of a valid INI file because we don't ever enter this check after the last \n is seen.